pkg/controller/metrics/provision_underway_collector_test.go

```diff
@@ -0,0 +1,519 @@
+			"cluster_deployment = cd-3 cluster_type = unspecified condition = DNSNotReady image_set = none namespace = cd-3 platform =  reason = FailedDueToQuotas 2",
+		},
+	}, {
+		name: "provisioning with no conditions and restarts less than min restarts",
+		existing: []runtime.Object{
+			cdBuilder("cd-1").Build(testcd.Installed()),
+			cdBuilder("cd-2").Build(testcd.InstallRestarts(1)),
+		},
+		min: 2,
+	}, {
+		name: "provisioning with other conditions and restarts less than min restarts",
+		existing: []runtime.Object{
+			cdBuilder("cd-1").Build(testcd.Installed()),
+			cdBuilder("cd-2").
+				Build(testcd.InstallRestarts(1), testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+					Type:   hivev1.ClusterHibernatingCondition,
+					Status: corev1.ConditionTrue,
+				})),
+		},
+		min: 2,
+	}, {
+		name: "provisioning with ProvisionFailed condition and restarts less than min restarts",
+		existing: []runtime.Object{
+			cdBuilder("cd-1").Build(testcd.Installed()),
+			cdBuilder("cd-2").
+				Build(testcd.InstallRestarts(1), testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+					Type:   hivev1.ProvisionFailedCondition,
+					Status: corev1.ConditionTrue,
+					Reason: "FailedDueToQuotas",
+				})),
+		},
+		min: 2,
+	}, {
+		name: "provisioning with ProvisionFailed, DNSNotReadyCondition condition and restarts less than min restarts",
+		existing: []runtime.Object{
+			cdBuilder("cd-1").Build(testcd.Installed()),
+			cdBuilder("cd-2").
+				Build(testcd.InstallRestarts(1), testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+					Type:   hivev1.ProvisionFailedCondition,
+					Status: corev1.ConditionTrue,
+					Reason: "FailedDueToQuotas",
+				})),
+			cdBuilder("cd-3").
+				Build(testcd.InstallRestarts(2), testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+					Type:   hivev1.DNSNotReadyCondition,
+					Status: corev1.ConditionTrue,
+					Reason: "FailedDueToQuotas",
+				})),
+		},
+		min: 2,
+		expected: []string{
+			"cluster_deployment = cd-3 cluster_type = unspecified condition = DNSNotReady image_set = none namespace = cd-3 platform =  reason = FailedDueToQuotas 2",
+		},
+	}}
+	for _, test := range cases {
+		t.Run(test.name, func(t *testing.T) {
+			c := fake.NewFakeClientWithScheme(scheme, test.existing...)
+			collect := newProvisioningUnderwayInstallRestartsCollectorWithMinRestarts(c, test.min)
+
+			descCh := make(chan *prometheus.Desc)
+			go func() {
+				for range descCh {
+				}
+			}()
+			collect.Describe(descCh)
+			close(descCh)
+			ch := make(chan prometheus.Metric)
+			go func() {
+				collect.Collect(ch)
+				close(ch)
+			}()
+
+			var got []string
+			for sample := range ch {
+				var d dto.Metric
+				require.NoError(t, sample.Write(&d))
+				got = append(got, metricPrettyWithValue(d))
+			}
+			assert.Equal(t, test.expected, got)
+		})
+	}
+}
```

Comment:
Bonus! I've read in some places not to unit test your metrics but I think in this case, with a custom collector, it makes sense.

Reply:
> I've read in some places not to unit test your metrics

it's code that can and will have assumptions, so unit tests make sense to make sure the assumptions don't change underneath you. So i don't get why we would not unit test collectors... unit tests also just help me understang how things work, _these are better than documentation in my case_ :P

Reply:
Yeah maybe their stance is more related to simpler instrumentation just reporting up a value. IIRC the reasoning was to not make it overly difficult to add metrics, so people would be more likely to do it.

Reply:
Unit tests are always a plus in my book